### PR TITLE
Field definition commands

### PR DESCRIPTION
**Pull request by Chi-teck**
This will bring [Field Definition Dumper](https://github.com/Chi-teck/field_definition_dumper) module into Drush core.

Ref: #5505

### CODE REVIEW

**Comment by Chi-teck** [comment]
I am not sure if `FieldDefinitionCommands` is a good name for these class. May be `FieldPluginCommands` or something else would work better.

File: src/Drupal/Commands/field/FieldDefinitionCommands.php [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+use Drush\Attributes as CLI;
+use Drush\Commands\DrushCommands;
+
+/**
+ * Defines field definition Drush commands.
+ */
+final class FieldDefinitionCommands extends DrushCommands {
+
+    public function __construct(
+        private readonly FieldTypePluginManagerInterface $typePluginManager,
```
Comment (weitzman) [inline src/Drupal/Commands/field/FieldDefinitionCommands.php:23]: Love the promoted properties. I used there on our Attribute classes - good idea to use in commandfiles.

File: src/Drupal/Commands/field/FieldDefinitionCommands.php [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+use Consolidation\OutputFormatters\StructuredData\RowsOfFields;
+use Drupal\Core\Field\BaseFieldDefinition;
+use Drupal\Core\Field\FieldTypePluginManagerInterface;
+use Drupal\Core\Field\FormatterPluginManager;
+use Drupal\Core\Field\WidgetPluginManager;
+use Drush\Attributes as CLI;
+use Drush\Commands\DrushCommands;
+
+/**
+ * Defines field definition Drush commands.
```
Comment (weitzman) [inline src/Drupal/Commands/field/FieldDefinitionCommands.php]: Lets omit useless comments. Our PHPCS doesnt require them AFAIK.

File: src/Drupal/Commands/field/FieldDefinitionCommands.php [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+    public function __construct(
+        private readonly FieldTypePluginManagerInterface $typePluginManager,
+        private readonly WidgetPluginManager $widgetPluginManager,
+        private readonly FormatterPluginManager $formatterPluginManager,
+    ) {
+        parent::__construct();
+    }
+
+    #[CLI\Command(name: 'field:types')]
```
Comment (weitzman) [inline src/Drupal/Commands/field/FieldDefinitionCommands.php]: Lets create a constant for each command name.

File: src/Drupal/Commands/field/FieldDefinitionCommands.php [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+            'id' => 'ID',
+            'label' => 'Label',
+            'default_widget' => 'Default Widget',
+            'default_formatter' => 'Default Formatter',
+            'settings' => 'Settings',
+        ],
+    )]
+    #[CLI\Help(
+        description: 'Lists field types.',
+        synopsis: 'drush field:types [options]',
```
Comment (weitzman) [inline src/Drupal/Commands/field/FieldDefinitionCommands.php]: Lets rely on Usages instead of this synopsis.

Comment (weitzman) [reply src/Drupal/Commands/field/FieldDefinitionCommands.php]: Also all other core commands use Doxygen for description and synopsis which is a bit more readable than Help attribute in my opinion and @greg-1-anderson opinion. Its a taste thing so OK to leave as is.

File: src/Drupal/Commands/field/FieldDefinitionCommands.php [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+            'provider' => 'Provider',
+        ],
+    )]
+    #[CLI\DefaultFields(fields: ['id', 'label', 'default_settings', 'field_types'])]
+    #[CLI\Help(
+        description: 'Lists field widgets.',
+        synopsis: 'drush field:widgets [options]',
+    )]
+    #[CLI\FilterDefaultField(field: 'id')]
+    public function widgets(array $options = ['format' => 'yaml', 'field-type' => NULL]): RowsOfFields {
```
Comment (weitzman) [inline src/Drupal/Commands/field/FieldDefinitionCommands.php]:
```suggestion
    public function widgets(array $options = ['format' => 'yaml', 'field-type' => self::REQ]): RowsOfFields {
```

File: src/Drupal/Commands/field/FieldDefinitionCommands.php [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+            'provider' => 'Provider',
+        ],
+    )]
+    #[CLI\DefaultFields(fields: ['id', 'label', 'default_settings', 'field_types'])]
+    #[CLI\Help(
+        description: 'Lists field formatters.',
+        synopsis: 'drush field:formatters [options]',
+    )]
+    #[CLI\FilterDefaultField(field: 'id')]
+    public function formatters(array $options = ['format' => 'yaml', 'field-type' => NULL]): RowsOfFields {
```
Comment (weitzman) [inline src/Drupal/Commands/field/FieldDefinitionCommands.php]:
```suggestion
    public function formatters(array $options = ['format' => 'yaml', 'field-type' => self::REQ]): RowsOfFields {
```

File: src/Drupal/Commands/field/FieldDefinitionCommands.php [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+        );
+    }
+
+    /**
+     * Encodes rows.
+     *
+     * Some output formats i.e. 'table' expect that each row to be a scalar
+     * value.
+     */
+    private static function encodeDefinitions(array $definitions, string $format): mixed {
```
Comment (weitzman) [inline src/Drupal/Commands/field/FieldDefinitionCommands.php]: See https://github.com/drush-ops/drush/blob/d013acfdb3fa0e48282e19640b065d584f9cb12e/src/Commands/core/StatusCommands.php#L87
https://github.com/weitzman/drush/blob/d013acfdb3fa0e48282e19640b065d584f9cb12e/src/Commands/core/StatusCommands.php#L87 for an approach we use in a few places for formatting table cells.

**Review by weitzman — COMMENTED** [review]
Thanks! Ideally we add some tests as well.

File: src/Drupal/Commands/field/FieldDefinitionCommands.php (hunk 7 above)
Comment (weitzman) [reply src/Drupal/Commands/field/FieldDefinitionCommands.php]: Or, do what https://github.com/drush-ops/drush/blob/d013acfdb3fa0e48282e19640b065d584f9cb12e/src/Commands/core/SiteCommands.php#L99
https://github.com/weitzman/drush/blob/d013acfdb3fa0e48282e19640b065d584f9cb12e/src/Commands/core/SiteCommands.php#L99 and others do and default to `format=yaml`. That handles nested data easily.

File: src/Drupal/Commands/field/FieldDefinitionCommands.php (hunk 4 above)
Comment (Chi-teck) [reply src/Drupal/Commands/field/FieldDefinitionCommands.php]: Replaced with Usage attribute.

File: src/Drupal/Commands/field/FieldDefinitionCommands.php (hunk 7 above)
Comment (Chi-teck) [reply src/Drupal/Commands/field/FieldDefinitionCommands.php]: That encoder is needed not only for table format but for some other formats that does not support nested structures. Seems like renderer function is not used for `string` format.

File: src/Drupal/Commands/field/FieldDefinitionCommands.php (hunk 3 above)
Comment (Chi-teck) [reply src/Drupal/Commands/field/FieldDefinitionCommands.php]: How should we name these constants?

Comment (weitzman) [reply src/Drupal/Commands/field/FieldDefinitionCommands.php]:
```
const TYPES =  'field:types'
```

File: src/Drupal/Commands/field/FieldDefinitionCommands.php (hunk 7 above)
Comment (weitzman) [reply src/Drupal/Commands/field/FieldDefinitionCommands.php]: I think all 3 commands should default to yaml format and return `UnstructuredListData` like https://github.com/weitzman/drush/blob/d013acfdb3fa0e48282e19640b065d584f9cb12e/src/Commands/core/SiteCommands.php#L99 . Then we wont need encodedefinitions() AFAICT.

Comment (weitzman) [reply src/Drupal/Commands/field/FieldDefinitionCommands.php]: I made a PR into this PR with my proposed changes https://github.com/Chi-teck/drush/pull/1

**Comment by weitzman** [comment]
I added some tests. Feedback welcome.

**Comment by Chi-teck** [comment]
Looks good to me.

**Comment by weitzman** [comment]
Tests are green. Anything else needed here @DieterHolvoet? If not feel free to merge.

**Comment by DieterHolvoet** [comment]
Looks good! Merged.

**Pull request merged**